src/modules/noit_lua_libnoit_binding.c

```diff
@@ -0,0 +1,263 @@
+  *addr = msg;
+  if(luaL_newmetatable(L, "metric_id_t") == 1) {
+    lua_pushcclosure(L, noit_metric_id_index_func, 0);
+    lua_setfield(L, -2, "__index");
+  }
+  lua_setmetatable(L, -2);
+}
+
+static int
+noit_message_index_func(lua_State *L) {
+  int n;
+  const char *k;
+  noit_metric_message_t **udata, *msg;
+  n = lua_gettop(L);    /* number of arguments */
+  mtevAssert(n == 2);
+  if(!luaL_checkudata(L, 1, "metric_message_t")) {
+    luaL_error(L, "metatable error, arg1 not a metric_message_t!");
+    return 1;
+  }
+  udata = lua_touserdata(L, 1);
+  msg = *udata;
+  if(!lua_isstring(L, 2)) {
+    luaL_error(L, "metatable error, arg2 not a string!");
+    return 1;
+  }
+
+  k = lua_tostring(L, 2);
+  switch (*k) {
+    case 'i':
+      if(!strcmp(k, "id")) {
+        noit_lua_setup_metric_id(L, &msg->id);
+      } else {
+        break;
+      }
+      return 1;
+    case 'w':
+      if(!strcmp(k, "whence_ms")) {
+        lua_pushinteger(L, msg->value.whence_ms);
+      } else {
+        break;
+      }
+      return 1;
+    case 'v':
+      if(!strcmp(k, "value")) {
+        lua_pushlightuserdata(L, (void*)&msg->value);
+      } else {
+        break;
+      }
+      return 1;
+    case 't':
+      if(!strcmp(k, "type")) {
+        lua_pushinteger(L, msg->type);
+      } else {
+        break;
+      }
+      return 1;
+    case 'o':
+      if(!strcmp(k, "original_message")) {
+        lua_pushstring(L, msg->original_message);
+      } else {
+        break;
+      }
+      return 1;
+    default:
+      break;
+  }
+  luaL_error(L, "metric_message_t no such element: %s", k);
+  return 0;
+}
+
+static int
+noit_lua_free_message(lua_State *L) {
+  const char *k;
+  noit_metric_message_t **udata;
+  if(!luaL_checkudata(L, 1, "metric_message_t")) {
+    luaL_error(L, "metatable error, arg1 not a metric_message_t!");
+    return 1;
+  }
+  udata = lua_touserdata(L, 1);
+
+  noit_metric_director_message_deref(*udata);
+  return 0;
+}
+
+static void
+noit_lua_setup_message(lua_State *L,
+    noit_metric_message_t *msg) {
+  noit_metric_message_t **addr;
+  addr = (noit_metric_message_t **)lua_newuserdata(L, sizeof(msg));
+  *addr = msg;
+  if(luaL_newmetatable(L, "metric_message_t") == 1) {
+    lua_pushcclosure(L, noit_message_index_func, 0);
+    lua_setfield(L, -2, "__index");
+
+    lua_pushcclosure(L, noit_lua_free_message, 0);
+    lua_setfield(L, -2, "__gc");
+  }
+  lua_setmetatable(L, -2);
+}
+
+static int
+lua_noit_metric_next(lua_State *L) {
+  noit_metric_message_t *msg;
+  msg = noit_metric_director_lane_next();
+  if(msg) {
+    noit_lua_setup_message(L, msg);
+    return 1;
+  }
+  return 0;
+}
+
+static int
+lua_noit_metric_messages_received(lua_State *L) {
+  lua_pushinteger(L, noit_metric_director_get_messages_received());
+  return 1;
+}
+
+static int
+lua_noit_metric_messages_distributed(lua_State *L) {
+  lua_pushinteger(L, noit_metric_director_get_messages_distributed());
+  return 1;
+}
+
+#ifndef NO_LUAOPEN_LIBNOIT
```

Comment:
I'd add a comment here directing the reader to the #define and #include at the top of `noit_lua_noit_binding.c`, to keep them aware of the dependency if they change this code. It is a simple dependency and the files are next to each other, but it would be an annoying issue to miss and have to debug.